src/Symfony/Component/Form/ChoiceList/ArrayChoiceList.php

```diff
@@ -207,4 +213,39 @@ protected function flatten(array $choices, $value, &$choicesByValues, &$keysByVa
             $structuredValues[$key] = $choiceValue;
         }
     }
+
+    /**
+     * Checks whether the given choices can be cast to strings without
+     * generating duplicates.
+     *
+     * @param array      $choices The choices.
+     * @param array|null $cache   The cache for previously checked entries. Internal
+     *
+     * @return bool Returns true if the choices can be cast to strings and
+     *              false otherwise.
+     */
+    private function castableToString(array $choices, array &$cache = null)
+    {
+        if (null === $cache) {
+            $cache = array();
+        }
+
+        foreach ($choices as $choice) {
+            if (is_array($choice)) {
+                if (!$this->castableToString($choice, $cache)) {
+                    return false;
+                }
+
+                continue;
+            } elseif (!is_scalar($choice)) {
+                return false;
+            } elseif (isset($cache[(string) $choice])) {
+                return false;
+            }
+
+            $cache[(string) $choice] = true;
```

Comment:
The cast does make a difference for the value `false`. By default, `false` is converted to `0` when adding it as array key. Here, it is converted to an empty string. I don't have a strong preference, but I lean towards converting `false` to an empty string.